numpy/array_api/linalg.py

```diff
@@ -0,0 +1,346 @@
+
+    # Note: this is different from np.linalg.pinv, which does not multiply the
+    # default tolerance by max(M, N).
+    if rtol is None:
+        rtol = max(x.shape[-2:]) * np.finfo(x.dtype).eps
+    return Array._new(np.linalg.pinv(x._array, rcond=rtol))
+
+def qr(x: Array, /, *, mode: Literal['reduced', 'complete'] = 'reduced') -> Tuple[Array, Array]:
+    """
+    Array API compatible wrapper for :py:func:`np.qr <numpy.qr>`.
+
+    See its docstring for more information.
+    """
+    # Note: qr is supposed to support stacks of matrices, but
+    # np.linalg.qr does not yet.
+
+    # Note: the restriction to floating-point dtypes only is different from
+    # np.linalg.qr.
+    if x.dtype not in _floating_dtypes:
+        raise TypeError('Only floating-point dtypes are allowed in qr')
+
+    # Note: the return type here is a namedtuple, which is different from
+    # np.linalg.qr, which only returns a tuple.
+    return qrresult(*map(Array._new, np.linalg.qr(x._array, mode=mode)))
+
+def slogdet(x: Array, /) -> slogdetresult:
+    """
+    Array API compatible wrapper for :py:func:`np.slogdet <numpy.slogdet>`.
+
+    See its docstring for more information.
+    """
+    # Note: the restriction to floating-point dtypes only is different from
+    # np.linalg.slogdet.
+    if x.dtype not in _floating_dtypes:
+        raise TypeError('Only floating-point dtypes are allowed in slogdet')
+
+    # Note: the return type here is a namedtuple, which is different from
+    # np.linalg.slogdet, which only returns a tuple.
+    return slogdetresult(*map(Array._new, np.linalg.slogdet(x._array)))
+
+def solve(x1: Array, x2: Array, /) -> Array:
+    """
+    Array API compatible wrapper for :py:func:`np.solve <numpy.solve>`.
+
+    See its docstring for more information.
+    """
+    # Note: the restriction to floating-point dtypes only is different from
+    # np.linalg.solve.
+    if x1.dtype not in _floating_dtypes or x2.dtype not in _floating_dtypes:
+        raise TypeError('Only floating-point dtypes are allowed in solve')
+
+    return Array._new(np.linalg.solve(x1._array, x2._array))
+
+def svd(x: Array, /, *, full_matrices: bool = True) -> svdresult:
+    """
+    Array API compatible wrapper for :py:func:`np.svd <numpy.svd>`.
+
+    See its docstring for more information.
+    """
+    # Note: the restriction to floating-point dtypes only is different from
+    # np.linalg.svd.
+    if x.dtype not in _floating_dtypes:
+        raise TypeError('Only floating-point dtypes are allowed in svd')
+
+    # Note: the return type here is a namedtuple, which is different from
+    # np.svd, which only returns a tuple.
+    return svdresult(*map(Array._new, np.linalg.svd(x._array, full_matrices=full_matrices)))
+
+# Note: svdvals is not in NumPy (but it is in SciPy). It is equivalent to
+# np.linalg.svd(compute_uv=False).
+def svdvals(x: Array, /) -> Union[Array, Tuple[Array, ...]]:
+    return Array._new(np.linalg.svd(x._array, compute_uv=False))
```

Comment:
```suggestion
def svdvals(x: Array, /) -> Array:
    return Array._new(np.linalg.svd(x._array, compute_uv=False))
```
This expression will never return a tuple of arrays.
I imagine this might be a leftover from an earlier version of the spec, one wherein `svd`'s keyword arguments were exposed?

Reply:
Yes, I think that was the case. I'll add this to https://github.com/data-apis/array-api/pull/277 as well.